### DISCUSSION TITLE
What am I missing to get the group management in the Wiki working?

### DISCUSSION THREAD

**Discussion by chrisanthropic** [CyberShadow/aconfmgr#127] [Q&A] [answered]
Hi,
I've really been enjoying getting my aconfmgr config setup. I'm nearly done but have been struggling to get the group managment function in the wiki to work.

More specifically, I've added the helper script suggested in the [User Management](https://github.com/CyberShadow/aconfmgr/wiki/User-management) section of the wiki and am successfully managing my `/etc/shadow` and `/etc/passwd` fully with aconfmgr and that function! If I modify the example below, aconfmgr will pick up the change and want to apply it.

Working `AddUser` example:
```
AddUser rtkit '!*' 133 133 x 'RealtimeKit' /proc /usr/bin/nologin ''
```

However, any attempts at calling `AddGroup` seem to go ignored. 

Failing `AddGroup` example:
```
AddGroup realtime x 967
```
Adding the above and running `aconfmgr apply` returns no changes, though when I check the `/tmp/aconfmgr-$USERNAME/groups` generated by the function, I can see a full copy of my existing `/etc/groups` and `/etc/gshadow` with the additional `realtime` group listed above added as expected.

I'd love to get my group management working alongside user management - can you help me figure out what I'm doing incorrectly?

**NOTE:** I've _also_ had to add `/etc/group` and `/etc/gpasswd` and `/etc/gshadow` to my ingore list. Otherwise, aconfmgr will _always_ attempt to copy it over, but the version it copies over doesn't include any changes reflected in the `/tmp/aconfmgr-$USERNAME/groups` .

**Comment by CyberShadow** [comment, marked as answer]
Sorry about that, looks like I forgot adding a helper when creating that page. Fixed that now:
https://github.com/CyberShadow/aconfmgr/wiki/User-management/_compare/7b5a5b15f2c1f4445353d23de71dedbade941847...f9839766e00144611df4af46edcf1db0d3217666

**Comment by chrisanthropic** [comment]
Thanks so much for the quick response and solution!

It's _mostly_ working now, but it's creating duplicate entries in both `/etc/group` and `/etc/gshadow`. (It also completely ignores `/etc/passwd`, I'm not sure if that was intentional or not though since that file is largely unused.)

For example:
```
AddUser rtkit '!*' 133 133 x 'RealtimeKit' /proc /usr/bin/nologin ''
AddGroup rtkit x 133
```
Results in this entry in both `/etc/group` and `/etc/gshadow`:
```
rtkit:x:133:
rtkit:x:133:
```

And I see this happen for any group I create that _also_ has a user of the same name.

**Reply by CyberShadow** [reply]
> It's _mostly_ working now, but it's creating duplicate entries in both `/etc/group` and `/etc/gshadow`.

Well, as the comment says, `AddUser` adds both the user and the corresponding group, as UNIX users generally always have an associated corresponding primary group. So, you don't need to call `AddGroup`, unless you're creating a group without an associated primary user (e.g. for special groups like `audio` or `rfkill`).

> It also completely ignores `/etc/passwd`, I'm not sure if that was intentional or not though since that file is largely unused.

Not sure why you would see that, though worth noting that `/etc/passwd` and `/etc/group` are [priority files](https://github.com/CyberShadow/aconfmgr/blob/56a51c921f386a6e02c42657544929b0544e82ba/src/common.bash#L68-L77), and are installed before any other changes if modified.

**Reply by chrisanthropic** [reply]
Ah, I see. I misunderstood. Thank you for the explanation and thanks again for the quick script update!

> It also completely ignores /etc/passwd, I'm not sure if that was intentional or not though since that file is largely unused.
That was a typo, sorry. I meant it's ignoring `/etc/gpaswd`.

**Reply by CyberShadow** [reply]
Arch Linux does not have an `/etc/gpasswd` file. The group-equivalent file of `/etc/passwd` is `/etc/group`. Or are you thinking of `/etc/gshadow`?

**Reply by chrisanthropic** [reply]
I was comparing the `aconfmgr` output on a brand new install to a backup of my previous install which is quite a few years old and from a time when Arch apparently still _did_ ship with using `/etc/gpasswd`.

Thanks again, apologies for the apparently unrelated questions.
